### PR TITLE
Added Disjoint Set Data Structure

### PR DESCRIPTION
**Pull request by NimrodRak**
I have added the Disjoint Set Data Structure implementation in C# using the Union By Rank and Path Compression heuristics.
- [x] I have performed a self-review of my code
- [x] My code follows the style guidelines of this project
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Comments in areas I changed are up to date
- [x] I have added comments to hard-to-understand areas of my code
- [x] I have made corresponding changes to the README.md

### CODE REVIEW

File: DataStructures.Tests/DisjointSet/DisjointSetTests.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        {
+            DisjointSet<int> ds = new DisjointSet<int>();
+            var one = ds.MakeSet(1);
+            var two = ds.MakeSet(2);
+            var three = ds.MakeSet(3);
+            ds.UnionSet(one, two);
+            Assert.IsTrue(ds.FindSet(one) == ds.FindSet(two));
+            ds.UnionSet(one, three);
+            Assert.IsTrue(ds.FindSet(two) == ds.FindSet(three));
+            Assert.IsTrue(one.Rank + two.Rank + three.Rank == 1);
```
Comment (siriak) [inline DataStructures.Tests/DisjointSet/DisjointSetTests.cs:27]: Please use FluentAssertions

File: DataStructures.Tests/DisjointSet/DisjointSetTests.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
+﻿using NUnit.Framework;
+
+namespace DataStructures.DisjointSet.Tests
+{
+    class DisjointSetTests
+    {
+        [Test]
+        public static void MakeTest()
```
Comment (siriak) [inline DataStructures.Tests/DisjointSet/DisjointSetTests.cs:8]: Please use descriptive names for tests

File: DataStructures.Tests/DisjointSet/DisjointSetTests.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
+﻿using NUnit.Framework;
+
+namespace DataStructures.DisjointSet.Tests
+{
+    class DisjointSetTests
```
Comment (siriak) [inline DataStructures.Tests/DisjointSet/DisjointSetTests.cs:5]: Please fix Codacy warnings

File: DataStructures/DisjointSet/Node.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+        public Node<T> Parent { get; set; }
+
+        public T Data { get; set; }
+
+        public Node(T data)
+        {
+            Parent = this;
+            Rank = 0;
+            this.Data = data;
```
Comment (siriak) [inline DataStructures/DisjointSet/Node.cs:19]:
```suggestion
        public Node<T> Parent { get; set; } = this;

        public T Data { get; set; }

        public Node(T data)
        {
            Data = data;
```

**Review by siriak — CHANGES_REQUESTED** [review]
Overall looks good, but I've left some comments

File: DataStructures.Tests/DisjointSet/DisjointSetTests.cs (hunk 1 above)
Comment (NimrodRak) [reply DataStructures.Tests/DisjointSet/DisjointSetTests.cs:27]: not NUnit? I saw the rest were with NUnit so i did the same.

**Comment by NimrodRak** [comment]
Fixed everything.

**Review by siriak — APPROVED** [review]
LGTM, thanks!

**Pull request merged**